From d9d91abe895f84220f078f597c8d5b00dc5bfa1f Mon Sep 17 00:00:00 2001
From: Peter McNeeley <petermcneeley@chromium.org>
Date: Tue, 21 Mar 2023 03:23:14 +0000
Subject: [PATCH] Optional mask filter in shared quad state mojo

The |MaskFilterInfo| member is rarely used and is moderately large
given that it contains a |RRect| object. Avoid serialization in mojo
empty. This only saves a few 10s of us per frame but there may be
other optimizations to follow.

Bug: 1396122
Change-Id: Ia86c50d48bab35860be66d8d5678514218293fd4
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4347778
Commit-Queue: Peter McNeeley <petermcneeley@chromium.org>
Reviewed-by: Joe Mason <joenotcharles@google.com>
Cr-Commit-Position: refs/heads/main@{#1119761}
---
 .../compositing/shared_quad_state_mojom_traits.h | 16 +++++++++++++---
 .../mojom/compositing/shared_quad_state.mojom    |  2 +-
 2 files changed, 14 insertions(+), 4 deletions(-)

--- a/services/viz/public/cpp/compositing/shared_quad_state_mojom_traits.h
+++ b/services/viz/public/cpp/compositing/shared_quad_state_mojom_traits.h
@@ -9,6 +9,7 @@
 #include "components/viz/common/quads/shared_quad_state.h"
 #include "services/viz/public/mojom/compositing/shared_quad_state.mojom-shared.h"
 #include "third_party/abseil-cpp/absl/types/optional.h"
+#include "ui/gfx/geometry/mask_filter_info.h"
 #include "ui/gfx/mojom/mask_filter_info_mojom_traits.h"
 #include "ui/gfx/mojom/rrect_f_mojom_traits.h"
 
@@ -38,9 +39,12 @@
     return input.sqs->visible_quad_layer_rect;
   }
 
-  static const gfx::MaskFilterInfo& mask_filter_info(
+  static const absl::optional<gfx::MaskFilterInfo> mask_filter_info(
       const OptSharedQuadState& input) {
-    return input.sqs->mask_filter_info;
+    return input.sqs->mask_filter_info.IsEmpty()
+               ? absl::nullopt
+               : absl::optional<gfx::MaskFilterInfo>(
+                     input.sqs->mask_filter_info);
   }
 
   static const absl::optional<gfx::Rect>& clip_rect(
@@ -118,11 +122,17 @@
     if (!data.ReadQuadToTargetTransform(&out->quad_to_target_transform) ||
         !data.ReadQuadLayerRect(&out->quad_layer_rect) ||
         !data.ReadVisibleQuadLayerRect(&out->visible_quad_layer_rect) ||
-        !data.ReadMaskFilterInfo(&out->mask_filter_info) ||
         !data.ReadClipRect(&out->clip_rect)) {
       return false;
     }
 
+    absl::optional<gfx::MaskFilterInfo> mask_filter;
+    if (!data.ReadMaskFilterInfo(&mask_filter)) {
+      return false;
+    }
+
+    out->mask_filter_info = mask_filter.value_or(gfx::MaskFilterInfo());
+
     out->are_contents_opaque = data.are_contents_opaque();
     out->opacity = data.opacity();
     if (data.blend_mode() > static_cast<int>(SkBlendMode::kLastMode))
--- a/services/viz/public/mojom/compositing/shared_quad_state.mojom
+++ b/services/viz/public/mojom/compositing/shared_quad_state.mojom
@@ -24,7 +24,7 @@
   // This rect lives in the target content space. It defines the mask filter
   // info applied to the quad, and also defines rounded corner rects to clip the
   // quads with.
-  gfx.mojom.MaskFilterInfo mask_filter_info;
+  gfx.mojom.MaskFilterInfo? mask_filter_info;
 
   // This rect lives in the target content space.
   gfx.mojom.Rect? clip_rect;
